builtin/game/item_entity.lua

```diff
@@ -152,6 +178,111 @@ core.register_entity(":__builtin:item", {
 			return
 		end
 
+		local node_this = core.get_node_or_nil(pos)
+		local is_stuck = false
+		if node_this then
+			local sdef = minetest.registered_nodes[node_this.name]
+			is_stuck = (sdef.walkable == nil or sdef.walkable == true)
+				and (sdef.collision_box == nil or sdef.collision_box.type == "regular")
+				and (sdef.node_box == nil or sdef.node_box.type == "regular")
+		end
+
+		-- Push item out when stuck inside solid node
+		if is_stuck then
+			local shootdir
+			local cx = (pos.x % 1) - 0.5
+			local cz = (pos.z % 1) - 0.5
+			local order = {}
+
+			-- First prepare the order in which the 4 sides are to be checked.
+			-- 1st: closest
+			-- 2nd: other direction
+			-- 3rd and 4th: other axis
+			local cxcz = function(o, cw, one, zero)
+				if cw < 0 then
+					table.insert(o, { [one]=1, y=0, [zero]=0 })
+					table.insert(o, { [one]=-1, y=0, [zero]=0 })
+				else
+					table.insert(o, { [one]=-1, y=0, [zero]=0 })
+					table.insert(o, { [one]=1, y=0, [zero]=0 })
+				end
+				return o
+			end
```

Comment:
You could set the other axis to 2 and 3 and the other direction to 4:
```lua
local function cxcz(o, sign, axis, also_other) -- What does that cxcz, cw, cx and co. stand for btw?
	local v = vector.new()
	v[axis] = sign
	table.insert(o, dir)
	if also_other then
		cxcz(o, -sign, axis, false)
	end
end

local order = {}
if math.abs(cx) > math.abs(cz) then
	cxcz(order, math.sign(cx), "x", false)
	cxcz(order, math.sign(cz), "z", true)
	cxcz(order, math.sign(cx), "x", false)
else
	-- ...
end
```

Reply:
LIke i wrote above, this is ridiculously overkill for just creating an order for looking at the 4 neighbours. Please don't waste time on this.